docs/src/type_stability.md

```diff
@@ -1,3 +1,219 @@
+
+```@example type-stability
+@code_warntype foo(1)
+```
+
+The key point is to ensure the return type of a function is clear from the types of its arguments. There are several ways to achieve this, and the best approach depends on the specific problem. For example, one can use the same return type:
+
+```@example type-stability
+function foo(x)
+    if x > 0
+        return 1.0
+    else
+        return -1.0
+    end
+end
+```
+
+Or you can ensure the return type matches the type of the argument:
+
+```@example type-stability
+function foo(x::T) where T
+    if x > 0
+        return T(1)
+    else
+        return -T(1)
+    end
+end
+```
+
+The latter example is very important because it takes advantage of Julia's multiple dispatch, which is one of the most powerful features of the language. Depending on the type `T` of the argument `x`, the Julia compiler generates a specialized version of `foo` that is optimized for this type. If the input type is an `Int64`, the return type is `Int64`, if `x` is a `Float64`, the return type is `Float64`, and so on.
+
+```@example type-stability
+@show foo(1)
+@show foo(-4.4)
+@show foo(1//2)
+nothing # hide
+```
+
+!!! note
+    If you didn't know how to make this function type-stable, it is probably a good idea to read the official Julia documentation, and in particular its [Performance Tips](https://docs.julialang.org/en/v1/manual/performance-tips/) section.
+
+## Global variables
+
+Another source of type instability is the use of global variables. In general, it is a good idea to declare global variables as `const` to ensure their type is fixed for the entire program. For example, consider the following function that internally takes a global variable `y`:
+
+```@example type-stability
+y = 2.4
+
+function bar(x)
+    res = zero(x) # this returns the zero of the same type of x
+    for i in 1:1000
+        res += y * x
+    end
+    return res
+end
+```
+
+The Julia compiler cannot infer the type of `res` because it depends on the type of `y`, which is a global variable that can change at any time of the program. We can check it using the `@code_warntype` macro:
+
+```@example type-stability
+@code_warntype bar(3.2)
+```
+
+While in the last example of the `foo` function we got a weak form of type instability, returning a `Union{Int, Float64}`, in this case the return type of `bar` is `Any`, meaning that the compiler doesn't know anything about the return type. Thus, this function has nothing different from a dynamically typed language like Python. We can benchmark the performance of `bar` using the [BenchmarkTools.jl](https://github.com/JuliaCI/BenchmarkTools.jl) package:
+
+```@example type-stability
+using BenchmarkTools
+
+@benchmark bar(3.2)
+```
+
+Here we see a lot of memory allocations and low performances in general. To fix this, we can declare a `const` variable instead:
```

Comment:
maybe add `const` (constant) ?
Just in case someone from Python doesn't understand the keyword (`const`)